kernel/src/utilities/single_thread_value.rs

```diff
@@ -0,0 +1,186 @@
+// 3. Simplicity for users vs. easy-to-make mistakes.
+//
+// Ideally, the guarantees this type provides would be verified at compile time
+// and be simple for users to use. However, as of July 2025, we don't know how
+// to realistically meet those goals.
+//
+// This approach first chooses to have automated enforcement rather than have
+// correctness (and avoiding unsoundness) based only on carefully written code
+// and scrutiny during code reviews. This is generally consistent with the Tock
+// ethos with using Rust and APIs to enforce correctness, even if the check
+// occurs at runtime.
+//
+// To implement the automated check, this type uses runtime checks. We don't
+// know a realistic way to move those checks to compile time.
+//
+// One downside to these decisions is the need to pass in the thread identifier
+// after the type is created. While this is conceptually similar to providing a
+// callback client after an object is constructed, which is widely used in
+// Tock, forgetting this operation will lead to the runtime check always
+// failing and will break the desired functionality of using the wrapped value.
+// However, passing in the type later is needed because the thread identifier
+// is likely not available where users want to construct this type.
+//
+// Ultimately, these trade-offs, runtime checks and an interface where missing a
+// call leads to failures, were deemed acceptable and worth the upside
+// (guaranteed soundness). In part, this is because this type should be used
+// and accessed sparingly within Tock. The complexity of creating this type
+// stems from its general risky-ness: enabling sharing static global variables.
+// While necessary for certain use cases within Tock, this should not be used
+// generally.
+pub struct SingleThreadValue<T> {
+    value: T,
+    thread_id: OptionalCell<usize>,
+    running_thread_id_fn: OptionalCell<fn() -> usize>,
+}
+
+impl<T> SingleThreadValue<T> {
+    /// Create a [`SingleThreadValue`].
+    ///
+    /// Note, the value will not be accessible immediately after `new()` runs.
+    /// To provide the single-thread guaranteed, [`SingleThreadValue`] needs a
+    /// reference to a [`ChipThreadId`] implementation, provided by
+    /// [`set_chip()`]. Since in many cases a suitable implementation of
+    /// [`ChipThreadId`] is not available when [`SingleThreadValue::new()`] is
+    /// called, the [`ChipThreadId`] implementation is provided later.
+    pub const fn new(value: T) -> Self {
+        Self {
+            value,
+            thread_id: OptionalCell::empty(),
+            running_thread_id_fn: OptionalCell::empty(),
+        }
+    }
+
+    /// Assign the [`ChipThreadId`] implementation.
+    ///
+    /// This stores the method that can identify the currently executing thread.
+    /// This method is used to determine if an attempted access is permitted or
+    /// not.
+    pub fn set_chip<C: crate::platform::chip::ChipThreadId>(&self) {
```

Comment:
I'm not sure why this is called `set_chip`? It's not _just_ setting the `thread_id`, but it's not setting the chip either. Maybe something like `initialize_thread()`?

Reply:
I agree, the naming in this PR is generally terrible. However, I'm wary of a name discussion not converging.

Reply:
The irony in that a type that's taken 1.5 years to create deserves careful naming while a type that's taken this long just needs to get created.